Load Data Local Infile

Hi Rene,
I'm using ProxySQL with Percona MySQL. I'm doing LOAD DATA LOCAL INFILE which works without ProxySQL but the same command when run through ProxySQL returns not found (Errcode: 2).

Please advise
thanks

Jenia

We are seeing a crash of the proxy related to `LOAD DATA LOCAL INFILE`

First there are a lot of errors logged where it says `File not found` which was to be expected but then we see a crash happening:

```
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
2019-09-05 14:10:13 MySQL_Session.cpp:2934:handler(): [WARNING] Error during query on (0,xx.xx.xx.xx,3306): 2, File '[somepath-to-a-file]' not found (Errcode: 2)
Error: signal 11:
/usr/bin/proxysql(_Z13crash_handleri+0x1a)[0x44c87a]
/lib64/libc.so.6(+0x36280)[0x7fbcc73b1280]
/usr/bin/proxysql[0x618156]
/usr/bin/proxysql(mysql_handle_local_infile+0x85)[0x6182f5]
/usr/bin/proxysql(mthd_my_read_query_result+0x85)[0x609fb5]
/usr/bin/proxysql[0x60f86f]
/usr/bin/proxysql(my_context_spawn+0x41)[0x60ec81]
2019-09-05 14:11:03 main.cpp:910:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2019-09-05 14:11:03 [INFO] Angel process started ProxySQL process 80629
```

The version used is 
```
# proxysql --version
ProxySQL version 1.4.14-percona-1.1, codename Truls
```

We've currently worked around it by blocking the `LOAD DATA LOCAL INFILE` queries using the query-rules to firewall out the these queries. But I wanted to alarm here that the implementation of the first stage of the solution seems not entirely bulletproof...

```
mysql> SELECT rule_id, active, apply, match_pattern, error_msg FROM mysql_query_rules;
+---------+--------+-------+-------------------------+--------------------------------------------------------------+
| rule_id | active | apply | match_pattern           | error_msg                                                    |
+---------+--------+-------+-------------------------+--------------------------------------------------------------+
| 1       | 1      | 1     | ^LOAD DATA LOCAL INFILE | LOAD DATA LOCAL INFILE commands are currently not supported! |
+---------+--------+-------+-------------------------+--------------------------------------------------------------+
1 row in set (0.00 sec)
```

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -1326,6 +1326,19 @@ bool MySQL_Session::handler_special_queries(PtrSize_t *pkt) {
 		l_free(pkt->size,pkt->ptr);
 		return true;
 	}
+	// 'LOAD DATA LOCAL INFILE' is unsupported. We report an specific error to inform clients about this fact. For more context see #833.
+	if ( (pkt->size >= 22 + 5) && (strncasecmp((char *)"LOAD DATA LOCAL INFILE",(char *)pkt->ptr+5, 22)==0) ) {
+		client_myds->DSS=STATE_QUERY_SENT_NET;
+		client_myds->myprot.generate_pkt_ERR(true,NULL,NULL,1,1047,(char *)"HY000",(char *)"Unsupported 'LOAD DATA LOCAL INFILE' command",true);
+		client_myds->DSS=STATE_SLEEP;
+		status=WAITING_CLIENT_DATA;
+		if (mirror==false) {
+			RequestEnd(NULL);
+		}
+		l_free(pkt->size,pkt->ptr);
+		return true;
+	}
+
 	return false;
 }
 

--- a/test/tap/tests/test_unsupported_queries-t.cpp
+++ b/test/tap/tests/test_unsupported_queries-t.cpp
@@ -1,0 +1,72 @@
+/**
+ * @file test_unsupported_queries-t.cpp
+ * @brief Simple test to check that unsupported queries by ProxySQL return the expected error codes.
+ */
+
+#include <cstring>
+#include <vector>
+#include <tuple>
+#include <string>
+#include <stdio.h>
+
+#include <mysql.h>
+#include <mysql/mysqld_error.h>
+
+#include "tap.h"
+#include "command_line.h"
+#include "utils.h"
+
+/**
+ * @brief List of the pairs holding the unsupported queries to be executed by ProxySQL
+ *   together with the error code that they should return.
+ */
+std::vector<std::tuple<std::string, int, std::string>> unsupported_queries {
+	std::make_tuple<std::string, int, std::string>("LOAD DATA LOCAL INFILE", 1047, "Unsupported 'LOAD DATA LOCAL INFILE' command"),
+	std::make_tuple<std::string, int, std::string>("LOAD DATA LOCAL INFILE 'data.txt' INTO TABLE db.test_table", 1047, "Unsupported 'LOAD DATA LOCAL INFILE' command"),
+	std::make_tuple<std::string, int, std::string>("LOAD DATA LOCAL INFILE '/tmp/test.txt' INTO TABLE test IGNORE 1 LINES", 1047, "Unsupported 'LOAD DATA LOCAL INFILE' command"),
+};
+
+int main(int argc, char** argv) {
+	CommandLine cl;
+
+	// plan as many tests as queries
+	plan(unsupported_queries.size());
+
+	if (cl.getEnv()) {
+		diag("Failed to get the required environmental variables.");
+		return -1;
+	}
+
+	// perform a different connection per query
+	for (const auto& unsupported_query : unsupported_queries) {
+		MYSQL* proxysql_mysql = mysql_init(NULL);
+
+		// extract the tuple elements
+		const std::string query = std::get<0>(unsupported_query);
+		const int exp_err_code = std::get<1>(unsupported_query);
+		const std::string exp_err_msg = std::get<2>(unsupported_query);
+
+		if (!mysql_real_connect(proxysql_mysql, cl.host, cl.username, cl.password, NULL, cl.port, NULL, 0)) {
+			fprintf(stderr, "File %s, line %d, Error: %s\n", __FILE__, __LINE__, mysql_error(proxysql_mysql));
+			return -1;
+		}
+
+		int query_err = mysql_query(proxysql_mysql, query.c_str());
+		int m_errno = mysql_errno(proxysql_mysql);
+		const char* m_error = mysql_error(proxysql_mysql);
+
+		ok(
+			query_err && ( m_errno == exp_err_code ) && ( exp_err_msg == std::string { m_error } ),
+			"Unsupported query '%s' should fail. Error code: (Expected: '%d' == Actual:'%d'), Error msg: (Expected: '%s' == Actual:'%s')",
+			query.c_str(),
+			exp_err_code,
+			m_errno,
+			exp_err_msg.c_str(),
+			m_error
+		);
+
+		mysql_close(proxysql_mysql);
+	}
+
+	return exit_status();
+}

